clang/lib/Sema/SemaChecking.cpp

```diff
@@ -439,6 +441,39 @@ static bool SemaBuiltinOverflow(Sema &S, CallExpr *TheCall,
     }
   }
 
+  // Mirror diagnostics from SemaExpr.cpp's diagnoseAmbiguousProvenance.
+  //
+  // ScalarExprEmitter::EmitSub's diagnostics aren't included here since
+  // they're generally unhelpful, grouped under pedantic warnings, and would be
+  // confusing without also taking into the type of the result.
+  const auto LHSExpr = TheCall->getArg(0);
+  const auto RHSExpr = TheCall->getArg(1);
+  if (BuiltinID != Builtin::BI__builtin_sub_overflow) {
+    assert((BuiltinID == Builtin::BI__builtin_add_overflow ||
+            BuiltinID == Builtin::BI__builtin_mul_overflow) &&
+           "Unexpected overflow builtin");
+
+    QualType LHSType = TheCall->getArg(0)->getType();
```

Comment:
Can we refactor diagnoseAmbiguousProvenance() and make it accessible to share this code? I think just passing Expr* instead of `ExprResult &` should work?